### PR TITLE
Implement faster Multinomial JAX dispatch

### LINKED ISSUE #868: Implement faster Multinomial JAX dispatch
### Description

We are defaulting to the numpyro implementation when that's installed, but the numpyro implementation is incredibly wasteful for large N, as it consists of doing n Categorical draws and summing up the values.

https://github.com/pyro-ppl/numpyro/blob/5af9ebda72bd7aeb08c61e4248ecd0d982473224/numpyro/distributions/util.py#L238

We should probably do sequential Binomial sampling: https://en.wikipedia.org/wiki/Multinomial_distribution#Algorithm:_Sequential_conditional_binomial_sampling

### PR DESCRIPTION
**Pull request by educhesne**
## Description
Replace the call to numpyro.distributions.util.multinomial by a custom function written in jax implementing [Sequential conditional binomial sampling](https://en.wikipedia.org/wiki/Multinomial_distribution#Algorithm:_Sequential_conditional_binomial_sampling)

## Related Issue

- [x] Closes #868 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/link/jax/dispatch/random.py [hunk 1]
```diff
@@ -450,6 +443,42 @@ def sample_fn(rng, size, dtype, n, p):
     return sample_fn
 
 
+def _jax_multinomial(n, p, shape=None, key=None):
+    if jnp.shape(n) != jnp.shape(p)[:-1]:
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/random.py:447]: Theres' some redundancy here. If `size` (not shape btw) is provided we should just broadcast `n` to `size` and `p` to `size + p.shape[-1]`. Only if `size` is not provided should we broadcast `n` and `p[...: -1]` together (basically finding the implicit size)

File: pytensor/link/jax/dispatch/random.py [hunk 2]
```diff
@@ -450,6 +443,42 @@ def sample_fn(rng, size, dtype, n, p):
...
+        p = jnp.broadcast_to(p, broadcast_shape + jnp.shape(p)[-1:])
+
+    if shape is not None:
+        broadcast_shape = jax.lax.broadcast_shapes(jnp.shape(n), shape)
+        n = jnp.broadcast_to(n, broadcast_shape)
+
+    else:
+        shape = p.shape[:-1]
+
+    p = p / jnp.sum(p, axis=-1, keepdims=True)
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/random.py:459]: We should assume p is valid, and not renormalize for the user. That would mask errors

File: pytensor/link/jax/dispatch/random.py [hunk 3]
```diff
@@ -450,6 +443,42 @@ def sample_fn(rng, size, dtype, n, p):
...
+    binom_p = jnp.moveaxis(p, -1, 0)[:-1, ...]
+
+    sampling_rng = jax.random.split(key, binom_p.shape[0])
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/random.py:462]: It seems you don't have enough rng splits here?, Unless you split it inside the scan as well? You will need a total of `(prod(size) *  binom_p.shape[0]` unique keys if I'm not mistaken.

Every binomial inside the scan needs a unique key, now it's reusing the same for every draw of the multinomial

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pytensor/link/jax/dispatch/random.py (hunk 1 above)
Comment (educhesne) [reply pytensor/link/jax/dispatch/random.py:447]: Yes, besides there is actually no need to broadcast p at all; `jax.random.binomial` broadcast it for us.
About the name `shape`: I kept the signature of the numpyro function but I can change it

Comment (ricardoV94) [reply pytensor/link/jax/dispatch/random.py:447]: you might be right about `p` although I guess it's more readable if you broadcast explicitly

File: pytensor/link/jax/dispatch/random.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/link/jax/dispatch/random.py:462]: Hmm the scan is taking all the draws simultaneously... In that case your split is probably correct, JAX will further split inside the binomial to draw each draw

File: pytensor/link/jax/dispatch/random.py [hunk 4]
```diff
@@ -450,6 +443,42 @@ def sample_fn(rng, size, dtype, n, p):
...
+        return ((s, rho), samples)
+
+    (remain, _), samples = jax.lax.scan(
+        _binomial_sample_fn,
+        (n.astype(np.float64), jnp.ones(binom_p.shape[1:])),
+        (binom_p, sampling_rng),
+    )
+    return jnp.concatenate(
+        [jnp.moveaxis(samples, 0, -1), jnp.expand_dims(remain, -1)], axis=-1
+    )
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/random.py:479]: I wonder if this is any faster than the useless binomial(p=1) draw if we didn't stop iterating before the last draw.

Comment (educhesne) [reply pytensor/link/jax/dispatch/random.py:479]: Just in case there are numerical approximations in the successive updates of p, here we ensure that the sum of samples is indeed n

Comment (ricardoV94) [reply pytensor/link/jax/dispatch/random.py:436]: you're right, sounds better this way

File: pytensor/link/jax/dispatch/random.py [hunk 5]
```diff
@@ -394,21 +395,47 @@ def sample_fn(rng_key, size, dtype, n, p):
...
-        sample = multinomial(key=rng_key, n=n, p=p, shape=size)
+        sample = _jax_multinomial(key=rng_key, n=n, p=p, size=size)
         return sample
 
     return sample_fn
 
 
+def _jax_multinomial(n, p, size=None, key=None):
+    if size is not None:
+        broadcast_shape_n = jax.lax.broadcast_shapes(jnp.shape(n), size)
```
[the remark this answers is not in the document]
Comment (educhesne) [reply pytensor/link/jax/dispatch/random.py:407]: Indeed, my bad

**Comment by educhesne** [comment]
It seems the `sample_fn` function is never called with `size=None` even though `pt.random.multinomial(..., size=None)` is in the test. Could it be that the `None` size is catched in between ? (there is a `static_shape` happening in `jax_funcify_RandomVariable` I don't fully understand)
Should I remove the part which is not reached by the coverage test ?

**Comment by ricardoV94** [comment]
It can be reached in some cases when for instance you have size=x.shape, and x is an input variable without known shape.

File: pytensor/link/jax/dispatch/random.py [hunk 6]
```diff
@@ -394,21 +395,44 @@ def sample_fn(rng_key, size, dtype, n, p):
...
-
     def sample_fn(rng_key, size, dtype, n, p):
-        sample = multinomial(key=rng_key, n=n, p=p, shape=size)
+        sample = _jax_multinomial(key=rng_key, n=n, p=p, size=size)
         return sample
 
     return sample_fn
 
 
+def _jax_multinomial(n, p, size=None, key=None):
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/random.py:405]: You can inline this in the dispatch for MultinomialRV. Also key=None is not really valid but we don't need default parameters anyway

**Comment by educhesne** [comment]
> It can be reached in some cases when for instance you have size=x.shape, and x is an input variable without known shape.

I'm sorry I don't know how to update the test to cover this case

Locally I've tried this without success:
```
p = np.array([[0.3, 0.7, 0.0], [0.1, 0.4, 0.5]])
size = (10_000, 2)
n = np.broadcast_to(np.array([10, 40]), size)
x = pt.matrix('x')

g = pt.random.multinomial(n, p, rng=rng, size=x.shape)
g_fn = compile_random_function([x], g, mode="JAX")
samples = g_fn(np.zeros(size))
```

**Comment by ricardoV94** [comment]
Sorry what I sad was non-sense. Here is a test with size=None

```python
import pytensor.tensor as pt

from tests.link.jax.test_random import compile_random_function

n = pt.vector("n")
p = pt.vector("p")

g = pt.random.multinomial(n, p, size=None)
g_fn = compile_random_function([n, p], g, mode="JAX")
```
Size is None, and it can't be inferred from `n` because `n` doesn't have a static shape.

**Review by ricardoV94 — APPROVED** [review]

**Comment by educhesne** [comment]
Thank you for your feedbacks !

**Comment by ricardoV94** [comment]
Thank you for the feature!!!
